[studio-ui] PathNavigators shared logic related refactors

aachavarria opened the pull request:
https://github.com/craftercms/craftercms/issues/4575

aachavarria committed:
Merge branch '4.0.0/develop' of https://github.com/rart/studio-ui into feature/4575

aachavarria committed:
#4575 Adding isPreviewable on node label click to display a preview dialog;

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -419,6 +425,46 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+
+  const onPreview = (item: DetailedItem) => {
+    if (item.systemType === 'component' || item.systemType === 'taxonomy') {
+      const src = `${legacyFormSrc}site=${site}&path=${item.path}&type=form&readonly=true`;
```
We should clean up all over moving this url construction to a single place. Possibly an epic or action builder or an util that one of those would invoke.

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -419,6 +425,46 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+        })
+      );
+    } else {
+      fetchContentXML(site, item.path).subscribe((content) => {
```
This here isn’t great either. 
Also, there’s no visibility of system status while loading the content.

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -419,6 +425,46 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+      fetchContentXML(site, item.path).subscribe((content) => {
+        let mode = 'txt';
+
+        if (item.systemType === 'renderingTemplate') {
```
All this mode selection logic needs to be encapsulated and made reusable.

aachavarria committed:
#4575 Creating getEditFormSrc to Create the Source used on LegacyFormDialog.tsx;
#4575 Updating all the files where showEditDialog is used to sent the src as individual parameters;

aachavarria committed:
#4575 Using Object.assign instead of spread on legacy files;
#4575 Fixes on QuickCreate, GlobalDialogManager and LegacyFormDialog.tsx to support new content source creation;

aachavarria committed:
#4575 Permissions renamed;

rart commented on static-assets/components/cstudio-contextual-nav/quick-create.js:
```diff
@@ -173,11 +172,12 @@ CStudioAuthoring.ContextualNav.WcmQuickCreate = CStudioAuthoring.ContextualNav.W
-          inProgress: false
-        });
+      const onQuickCreateItemSelected = (props) => {
+        showEditDialog(
+          Object.assign(props, {
+            inProgress: false
+          })
+        );
```
The changes don't look compatible. Previously receiving a `src` (string?) and now an object `props`. I don't see the counter part of this change. Was this broken?

rart commented on ui/app/src/components/PathNavigator/utils.ts:
```diff
@@ -17,23 +17,41 @@
+
+export function isEditableViaFormEditor(item: DetailedItem | SandboxItem): boolean {
+  return item.systemType === 'component' || item.systemType === 'taxonomy';
+}
```
Problem is this function is lying because pages are editable via form too.

rart commented on ui/app/src/modules/Preview/Tools/EditFormPanel.tsx:
```diff
@@ -189,12 +189,38 @@ function EditFormPanelBody(props: EditFormPanelBodyProps) {
+        modelId: selectedId
+      };
+    }
+  }, [authoringBase, childrenMap, model, models, path, selectedId, site]);
```
Doesn't look like you're using this logic elsewhere. Why not do this directly inside openDialog?

rart commented on ui/app/src/utils/itemActions.ts:
```diff
@@ -726,12 +728,14 @@ export const itemActionDispatcher = ({
-        let src = `${legacyFormSrc}site=${site}&path=${encodeURIComponent(item.path)}&type=asset&readonly=true`;
+        let src = `${authoringBase}/legacy/form?site=${site}&path=${encodeURIComponent(
+          item.path
+        )}&type=asset&readonly=true`;
```
Shouldn't these use `getEditFormSrc`?

rart commented on ui/app/src/utils/path.ts:
```diff
@@ -144,3 +145,38 @@ export function getPasteItemFromPath(path: string, paths: string[]): PasteItem {
+  contentTypeId?: string;
+  isNewContent?: boolean;
+}): string {
+  const qs = toQueryString({
+    site,
+    path,
+    type: 'form',
+    readonly,
+    isHidden,
+    modelId,
+    changeTemplate,
+    contentTypeId,
+    isNewContent
+  });
```
Make sure query-string lib excludes undefined props

aachavarria committed:
#4575 PR Feedback;

aachavarria committed:
Merge branch 'develop' of https://github.com/craftercms/studio-ui into feature/4575

aachavarria committed:
#4575 Removing wrong code

aachavarria committed:
#4575 showCodeEditorDialog refactor;

aachavarria committed:
#4575 Adding Refresh feature to pathNavigatorTree.ts;
#4575 Adding deleted Cut Event Propagation;

aachavarria committed:
#4575 Adding TODO and deleting console.log;

aachavarria committed:
[4575] Error 404 issue handle;
[4575] DeleteItems API Changes to receive a array of paths;

aachavarria committed:
[4575] deleting console.log;

aachavarria committed:
[4575] Creating new action, reducer, epic for pathNavigatorTreeFetchPathsChildren;

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -205,7 +216,7 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
   useEffect(() => {
     keywordByPathRef.current = keywordByPath;
-  }, [keywordByPath]);
+  }, [keywordByPath, itemsByPath]);
```
`itemsByPath` not used on effect. Why is it listed as a dependency?

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -241,8 +252,33 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+          if (payload.clipboard?.type === 'CUT') {
+            const sourceNode =
+              nodesByPathRef.current[getParentPath(payload.clipboard.sourcePath)] ??
+              nodesByPathRef.current[withIndex(getParentPath(payload.clipboard.sourcePath))];
```
extract `getParentPath` call to a const to avoid double computation

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -241,8 +252,33 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
         case folderCreated.type: {
-          if (payload.clipboard.type === 'CUT') {
-            debugger;
+          if (payload.clipboard?.type === 'CUT') {
```
Confirm: it should go to else statement whenever is not a cut **including** paste items via copy?
Otherwise, this condition needs to be revisited and rather use `type` to make the determination.

aachavarria replied:
that is ok, it should go to the else statement for any operation excluding CUT, in this case COPY or folderCreated

rart commented on ui/app/src/components/PathNavigatorTree/PathNavigatorTree.tsx:
```diff
@@ -241,8 +252,33 @@ export default function PathNavigatorTree(props: PathNavigatorTreeProps) {
+              nodesByPathRef.current[getParentPath(payload.clipboard.sourcePath)] ??
+              nodesByPathRef.current[withIndex(getParentPath(payload.clipboard.sourcePath))];
+            const targetNode =
+              nodesByPathRef.current[payload.target] ?? nodesByPathRef.current[withIndex(payload.target)];
```
Should we be doing rather this:
```suggestion
              nodesByPathRef.current[withoutIndex(payload.target)] ?? nodesByPathRef.current[withIndex(payload.target)];
```
We don't know if an items is registered with or without index on the state so if not using both `withoutIndex` and `withIndex` we might be checking for the same thing and not getting the result we want.

We should make an util that does this
```ts
export function lookupItemByPath(path: string, lookupTable: LookupTable<Sandbox>): Sandbox
export function lookupItemByPath(path: string, lookupTable: LookupTable<Detailed>): Detailed 
export function lookupItemByPath(path: string, lookupTable: LookupTable<Sandbox | Detailed>): Sandbox | Detailed {
  return lookupTable[withoutIndex(path)] ?? lookupTable[withInde(path)];
}
```

Same deal would apply above to `sourceNode`. See if there's anywhere else that could use the new util.

aachavarria committed:
[4575] Creating util lookupNodeByPath, PR Feedback;

rart commented on ui/app/src/components/PathNavigatorTree/utils.ts:
```diff
@@ -0,0 +1,23 @@
+
+export function lookupNodeByPath(path: string, lookupTable: LookupTable<TreeNode>): TreeNode {
+  return lookupTable[withoutIndex(path)] ?? lookupTable[withIndex(path)];
+}
```
```suggestion
export function lookupItemByPath<T = DetailedItem>(path: string, lookupTable: LookupTable<T>): T {
  return lookupTable[withoutIndex(path)] ?? lookupTable[withIndex(path)];
}
```

rart replied:
This should be on utils/content

aachavarria committed:
[4575] PR Feedback;
